File: services/agent/src/usageQueue.ts
export interface UsageRecord {
  action: string;
  chars_processed: number;
  credits_spent: number;
  timestamp?: string;
}

export class UsageQueue {
  private queue: UsageRecord[] = [];

  enqueue(record: Omit<UsageRecord, 'timestamp'>) {
    this.queue.push({ ...record, timestamp: new Date().toISOString() });
  }

  flush(): UsageRecord[] {
    const records = [...this.queue];
    this.queue = [];
    return records;
  }

  size(): number { return this.queue.length; }
}

export const usageQueue = new UsageQueue();

File: services/agent-rs/src/usage_queue.rs
use chrono::Utc;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UsageRecord {
    pub action:          String,
    pub chars_processed: u64,
    pub credits_spent:   i64,
    pub timestamp:       String,
}

#[derive(Default)]
pub struct UsageQueue {
    records: Vec<UsageRecord>,
}

impl UsageQueue {
    pub fn new() -> Self { Self::default() }

    pub fn enqueue(&mut self, action: String, chars_processed: u64, credits_spent: i64) {
        self.records.push(UsageRecord {
            action,
            chars_processed,
            credits_spent,
            timestamp: Utc::now().to_rfc3339(),
        });
    }

    pub fn flush(&mut self) -> Vec<UsageRecord> {
        std::mem::take(&mut self.records)
    }

    pub fn size(&self) -> usize { self.records.len() }
}
